proposals/distribution.md

```diff
@@ -103,19 +103,17 @@ The following entries should be added to the [scope table][scope]:
 * “Specifying a distribution method”.
     This entry replaces part of the previous “Creating Reference spec for optional DNS based naming & distribution” and “Standardizing on a particular Distribution method” entries.
 
-    Retrieving image indexes covers the current “tag listing” (e.g. “what named manifests are in `library/busybox`?”), because tags are entries in the image format's [`manifests` array][manifests].
-    Other tag-listing endpoints needed for backwards-compatibility are therefore in scope as well.
+    Tag-listing (e.g. “what named manifests are in `library/busybox`?”) endpoints are in scope and required for backwards compatibility with clients.
+    External methods for listing tags remain compatible through fetching by content address, but are out of scope.
```

Comment:
> I can guess that this refers to using an image index to approximate listing tags

What do you mean by this? I was thinking an external method would completely replace the tag listing, rather than just approximate it. There is a gap in the existing API where manifests must be uploaded with a tag, but we can easily address that especially for cases where tagging is not used at all within a registry. Since something like garbage collection does not belong in this specification, having a requirement about the relationship between tags and content is not required, allowing tags to be completely external.

Reply:
> What do you mean by this?

This PR drops `Retrieving image indexes covers the current “tag listing”` and adds `External methods for listing tags`, so I assumed you replaced the former with the latter to be more general?

I don't really see the need to mention this. It seems irrelevant, but I might just not understand what you're trying to say.

> There is a gap in the existing API where manifests must be uploaded with a tag

[Not per the spec](https://github.com/docker/distribution/blob/5cb406d511b7b9163bff9b6439072e4892e5ae3b/docs/spec/api.md#put-manifest). IIRC, docker doesn't allow you to push by a digest (maybe DockerHub too?), but I know at least GCR allows it.

> Since something like garbage collection does not belong in this specification, having a requirement about the relationship between tags and content is not required, allowing tags to be completely external.

I don't see how tags can be completely external if tag listing is part of the spec. Are you saying tags are (mostly) undefined behavior in order to give freedom to implementors?

Reply:
> I don't really see the need to mention this. It seems irrelevant, but I might just not understand what you're trying to say.

I was mostly just trying to account for the language which was here before, it really doesn't need to be mentioned. I will remove it, the scope section below is already clear that specification supports managing content addressable objects, what clients end up doing with that doesn't need to be mentioned here.

> Not per the spec

Fair enough, the spec is clear that references can be digests, the incompatibility was mostly related to the deprecated schema 1 I believe. We will clean up these examples in the specification once it is in OCI.

> I don't see how tags can be completely external if tag listing is part of the spec

I am only saying that a registry should be usable as only a CAS, without tagging at all. In which case any sort of resolution of a string -> content address is the responsibility of the distribution client. This does provide more freedom to distribution clients.

Reply:
> just trying to account for the language which was here before

Makes sense! I appreciate your patience :)

> I am only saying that a registry should be usable as only a CAS, without tagging at all.

Right, I totally agree. My worry is that a registry that only implements the CAS portions of the spec will be seen as non-conforming.

> resolution of a string -> content address is the responsibility of the distribution client

I think this is my ideal. Tag listing and tag resolution seem to be more about discovery/naming than distribution, and registries that wish to be compatible with older clients could choose to implement these via the old tags list endpoint.

Regardless, I think this is a good starting point. I just hope tag listing being a requirement isn't a foregone conclusion.